### PR TITLE
Start removing event system

### PR DESCRIPTION
**Pull request by vchuravy**
As I have mentioned over the last little while I would like to remove the event
system. Over the years it has proven itself to be cumbersome, and prone to introduce
bugs. 

In contrast @maleadt developed a task-based concurrency scheme for CUDA.jl that is rather
elegant and doesn't introduce a new terminology beside Julia's existing tasks.

This is the last big chance I forsee and I plan on KernelAbstractions 0.9 to be a direct
pre-cursor to 1.0.

Things that need to be hammered out:

1. Does this also work with the oneAPI and ROCm backend?
2. `synchronize(::Device)` is rather nice, but is also exported by CUDA.jl
3. For 1.0 can we add KA as a dependency to GPUArrays.jl and move CUDAKernels.jl into CUDA.jl

### CODE REVIEW

File: docs/src/quickstart.md [hunk 1]
```diff
@@ -87,29 +86,22 @@ using CUDAKernels # Required to access CUDADevice
...
 ```julia
 using CUDAKernels # Required to access CUDADevice
 function mymul(A::CuArray, B::CuArray)
     A .= 1.0
     B .= 3.0
-    evA = mul2_kernel(CUDADevice(), 16)(A, ndrange=size(A), dependencies=Event(CUDADevice()))
-    evB = mul2_kernel(CUDADevice(), 16)(A, ndrange=size(A), dependencies=Event(CUDADevice()))
-    wait(evA, evB)
+    mul2_kernel(CUDADevice(), 16)(A, ndrange=size(A))
+    mul2_kernel(CUDADevice(), 16)(A, ndrange=size(A))
```
Comment (vchuravy) [inline docs/src/quickstart.md:100]:
```suggestion
    mul2_kernel(CUDADevice(), 16)(A, ndrange=size(A))
    synchronize(CUDADevice())
```

File: examples/matmul.jl [hunk 2]
```diff
@@ -36,7 +36,7 @@ c = zeros(256, 45)
 
 # beginning CPU tests, returns event
 ev = matmul!(a,b,c)
```
Comment (vchuravy) [inline examples/matmul.jl:38]:
```suggestion
matmul!(a,b,c)
```

File: examples/matmul.jl [hunk 3]
```diff
@@ -36,7 +36,7 @@ c = zeros(256, 45)
...
-wait(ev)
+synchronize(device)
```
Comment (vchuravy) [inline examples/matmul.jl:39]:
```suggestion
synchronize(KernelAbstractions.get_device(a))
```

File: examples/matmul.jl [hunk 4]
```diff
@@ -47,7 +47,7 @@ if has_cuda && has_cuda_gpu()
     d_c = CuArray(c)
 
     ev = matmul!(d_a, d_b, d_c)
-    wait(ev)
+    synchronize(device)
```
Comment (vchuravy) [inline examples/matmul.jl:50]:
```suggestion
    synchronize(KernelAbstractions.get_device(a))
```

File: examples/memcopy.jl [hunk 5]
```diff
@@ -14,8 +14,8 @@ end
 
 A = zeros(128, 128)
 B = ones(128, 128)
-event = mycopy!(A, B)
-wait(event)
+mycopy!(A, B)
+synchronize(device)
```
Comment (vchuravy) [inline examples/memcopy.jl:18]:
```suggestion
synchronize(KernelAbstractions.get_device(A))
```

File: examples/memcopy.jl [hunk 6]
```diff
@@ -27,8 +27,8 @@ if has_cuda && has_cuda_gpu()
 
     A = CuArray{Float32}(undef, 1024)
     B = CUDA.ones(Float32, 1024)
-    event = mycopy!(A, B)
-    wait(event)
+    mycopy!(A, B)
+    synchronize(device)
```
Comment (vchuravy) [inline examples/memcopy.jl:31]:
```suggestion
    synchronize(KernelAbstractions.get_device(A))
```

File: examples/memcopy.jl [hunk 7]
```diff
@@ -40,7 +40,7 @@ if has_rocm && has_rocm_gpu()
 
     A = zeros(Float32, 1024) |> ROCArray
     B = ones(Float32, 1024) |> ROCArray
-    event = mycopy!(A, B)
-    wait(event)
+    mycopy!(A, B)
+    synchronize(device)
```
Comment (vchuravy) [inline examples/memcopy.jl:44]:
```suggestion
    synchronize(KernelAbstractions.get_device(A))
```

File: examples/memcopy_static.jl [hunk 8]
```diff
@@ -28,7 +28,7 @@ if has_cuda && has_cuda_gpu()
 
     A = CuArray{Float32}(undef, 1024)
     B = CUDA.ones(Float32, 1024)
-    event = mycopy_static!(A, B)
-    wait(event)
+    mycopy_static!(A, B)
+    synchronize(CPU())
```
Comment (vchuravy) [inline examples/memcopy_static.jl:32]:
```suggestion
    synchronize(KernelAbstractions.get_device(a))
```

File: examples/memcopy_static.jl [hunk 9]
```diff
@@ -14,8 +14,8 @@ end
 
 A = zeros(128, 128)
 B = ones(128, 128)
-event = mycopy_static!(A, B)
-wait(event)
+mycopy_static!(A, B)
+synchronize(CPU())
```
Comment (vchuravy) [inline examples/memcopy_static.jl:18]:
```suggestion
synchronize(KernelAbstractions.get_device(A))
```

File: src/cpu.jl [hunk 10]
```diff
@@ -1,95 +1,37 @@
...
-    if any(failed, events)
-        ex = CompositeException()
-        for event in events
-            if failed(event) && event isa CPUEvent
-                push!(ex, Base.TaskFailedException(event.task))
-            end
-        end
-        throw(ex)
+function synchronize(::CPU)
+    last = get(task_local_storage, :KA_CPU, nothing)
```
Comment (vchuravy) [inline src/cpu.jl:4]:
```suggestion
    last = Base.get(task_local_storage, :KA_CPU, nothing)
```

File: src/cpu.jl [hunk 11]
```diff
@@ -1,95 +1,37 @@
...
-    if progress === nothing
-        wait(ev.task)
-    else
-        while !Base.istaskdone(ev.task)
-            progress()
-        end
-        if Base.istaskfailed(ev.task)
-            throw(Base.TaskFailedException(ev.task))
+@inline function do_async(f::F, args...; kwargs...) where F
+    last = get(task_local_storage(), :KA_CPU, nothing)
```
Comment (vchuravy) [inline src/cpu.jl:11]:
```suggestion
    last = Base.get(task_local_storage(), :KA_CPU, nothing)
```

File: src/cpu.jl [hunk 12]
```diff
@@ -1,95 +1,37 @@
...
-    if any(failed, events)
-        ex = CompositeException()
-        for event in events
-            if failed(event) && event isa CPUEvent
-                push!(ex, Base.TaskFailedException(event.task))
-            end
-        end
-        throw(ex)
+function synchronize(::CPU)
+    last = Base.get(task_local_storage, :KA_CPU, nothing)
```
Comment (vchuravy) [inline src/cpu.jl:4]:
```suggestion
    last = Base.get(task_local_storage(), :KA_CPU, nothing)
```

File: lib/CUDAKernels/src/CUDAKernels.jl [hunk 13]
```diff
@@ -209,22 +37,16 @@ function __pin!(a)
     return nothing
 end
 
-function KernelAbstractions.async_copy!(::CUDADevice, A, B; dependencies=nothing, progress=yield)
+function KernelAbstractions.async_copy!(::CUDADevice, A, B, progress=yield)
```
Comment (lcw) [inline lib/CUDAKernels/src/CUDAKernels.jl:40]: Is the `progress` argument still needed?

**Comment by leios** [comment]
We recently had a discussion about using KernelAbstractions in GPUArrays. To do this, we need to:
1. Fix up AMDGPU to work with this PR
2. Move CUDAKernels, ROCKernels, etc to AMDGPU, CUDA, etc... (as package extensions in 1.9?)
3. Make GPUArrays dependent on KA

**Pull request merged**